src/04-extensions.js

```diff
@@ -0,0 +1,212 @@
+      "WHQcpcOj0G",
+      "GKl035EYKN",
+      "2GglUqKT0G",
+      "Pr6kc4Q_Xf",
+      "ft5Gs5izdq",
+      "wuL4ddBinQ",
+    ],
+    // ...
+  }
+*/
+
+function organizeDinosaursByPeriod(dinos, key) {
+  //input an array of objects
+  //output an object where key represents period and value reps an array of dinoIDs
+
+  //create helper function to camelCase
+  function camelCaseKey(keyString) {
+    //create an array that contains all valid keys
+    const validKeys = ["name", "pronunciation", "meaningOfName", "diet", "lengthInMeters", "mya", "info"];
+
+    //iterate through validKeys and if the keyString is included in array then assign it
+    for (let i = 0; i < validKeys.length; i++) {
+      if (validKeys[i].toLowerCase() === keyString.toLowerCase()) {
+        keyString = validKeys[i];
+        break;
+      }
+    }
+
+    return keyString;
+  }
+
+  //if key is undefined assign it the string dinosaurId
+  if (key === undefined) key = "dinosaurId";
+  else key = camelCaseKey(key);
+  //edgecase - all dinos share same keys, so check whether key exists in object
+  if (!dinos[0][key] || key === "period") return "please provide a valid key: (name,pronunciation,meaningOfName,diet,lengthInMeters,mya, or info)";
+
+  //initialie an empty object
+  const result = {};
+
+  //iterate through dinosaurs
+  dinos.forEach((dino) => {
+    //if period already exists in object, push dinoID to object at key
+    if (result[dino.period]) result[dino.period].push(dino[key]);
+    //otherwise create key in object and assign it array with current dinoID in it
+    else result[dino.period] = [dino[key]];
+  });
+
+  return result;
+}
+
+const result = organizeDinosaursByPeriod(exampleDinosaurData, "mya");
+// console.log(result);
+
+//--SPACING
+
+/*
+Write a function that takes in an array of strings, where each string is a roomId. Assume that someone is attempting to go from room to 
+room, starting with the room at index 0 and then preceding to the room at index 1, and so on.
+Return true if the path the person is trying to follow is possible. Otherwise, return false.
+*/
+
+const valid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "0eNtkY5WoA"];
+const invalid = ["aIA6tevTne", "A6QaYdyKra", "L72moIRcrX", "dpQnu5wgaN"];
+
+function validateMuseumPath(selectedPath, showPath) {
+  //input: array of strings that represents roomIDs and specific order a user wishes to go room to room
+  //output: a boolean value indicating whether the path is possible or not
+
+  //edge-case: empty array
+  if (selectedPath.length === 0) return "array cannot be empty";
+
+  //declare a lastRoom variable and set to null originally
+  let lastRoom = null;
+  //make copy of exampleRoomData
+  remainingRooms = exampleRoomData.slice();
+
+  //make sure they start off at coat check room
+  if (selectedPath[0] !== remainingRooms[2].roomId) {
+    return "path must start at 'aIA6tevTne' a.k.a Coat Check Room";
+  } else {
+    lastRoom = exampleRoomData[2].roomId;
```

Comment:
Because you're using `exampleRoomData` here, this function now depends on global scope. This is probably my bad though, I should've said to include `rooms` as a parameter. 😅